tests/testsuite.src/syn_misc.at

```diff
@@ -8348,3 +8348,258 @@ AT_CHECK([$COBCRUN_DIRECT ./prog], [0], [1])
+       PROGRAM-ID. prog4.
+       DATA DIVISION.
+       LOCAL-STORAGE SECTION.
+       PROCEDURE DIVISION.
+          DISPLAY "main".
+          CALL "entry".
+	  DISPLAY "end-main".
+          GOBACK.
+          ENTRY "entry".
+          DISPLAY "entry".
+])
+
+AT_CHECK([$COMPILE -x prog4.cob])
+
+AT_CHECK([./prog4], [0], [main
+entry
+end-main
+])
+
+AT_DATA([prog5.cob], [
+       IDENTIFICATION DIVISION.
+       PROGRAM-ID. prog5 RECURSIVE.
+       DATA DIVISION.
+       WORKING-STORAGE SECTION.
+       01 counter PIC 9(1).
+       PROCEDURE DIVISION.
+          DISPLAY "HELLO WORLD".
+          CALL "entry".
+          GOBACK.
+          ENTRY "entry".
+	  ADD 1 TO counter.
+	  DISPLAY "entry " counter.
+	  IF counter < 5 THEN
+            CALL "inside-program"
+          END-IF.
+	  GOBACK.
+       PROGRAM-ID. inside-program RECURSIVE.
+       PROCEDURE DIVISION.
+           CALL "entry".
+       END PROGRAM inside-program.
+       END PROGRAM prog5.
+])
+
+# There is no reason for this one to fail, it has the RECURSIVE flag...
+# but no IDENTIFICATION DIVISION before the second PROGRAM-ID
+
+AT_CHECK([$COMPILE -x prog5.cob], [1], [],
+[prog5.cob:18: error: syntax error, unexpected Identifier
+])
+AT_DATA([prog5.cob], [
+       IDENTIFICATION DIVISION.
+       PROGRAM-ID. prog5 RECURSIVE.
+       DATA DIVISION.
+       WORKING-STORAGE SECTION.
+       01 counter PIC 9(1).
+       PROCEDURE DIVISION.
+          DISPLAY "HELLO WORLD".
+          CALL "entry".
+          GOBACK.
+          ENTRY "entry".
+	  ADD 1 TO counter.
+	  DISPLAY "entry " counter.
+	  IF counter < 5 THEN
+            CALL "inside-program"
+          END-IF.
+	  GOBACK.
+       IDENTIFICATION DIVISION.
+       PROGRAM-ID. inside-program RECURSIVE.
+       PROCEDURE DIVISION.
+           CALL "entry".
+       END PROGRAM inside-program.
+       END PROGRAM prog5.
+])
+
+# Adding the IDENTIFICATION DIVISION makes it work... why ?
+
+AT_CHECK([$COMPILE -x prog5.cob])
+
+AT_CHECK([./prog5], [0], [HELLO WORLD
+entry 1
+entry 2
+entry 3
+entry 4
+entry 5
+])
+
+AT_DATA([prog6.cob], [
+       IDENTIFICATION DIVISION.
+       PROGRAM-ID. prog6 RECURSIVE.
+       DATA DIVISION.
+       WORKING-STORAGE SECTION.
+       01 counter PIC 9(1).
+       PROCEDURE DIVISION.
+          DISPLAY "HELLO WORLD".
+          CALL "entry".
+          GOBACK.
+          ENTRY "entry".
+	  ADD 1 TO counter.
+	  DISPLAY "entry " counter.
+	  IF counter < 5 THEN
+            CALL "inside-program"
+          END-IF.
+	  GOBACK.
+       IDENTIFICATION DIVISION.
+       PROGRAM-ID. inside-program.
+       DATA DIVISION.
+       LOCAL-STORAGE SECTION.
+       PROCEDURE DIVISION.
+           CALL "entry".
+       END PROGRAM inside-program.
+       END PROGRAM prog6.
+])
+
+# No LOCAL-STORAGE in nested programs ? Microfocus documentation does
+# not display such a limitation
+
+AT_CHECK([$COMPILE -x prog6.cob], [1], [],
+[prog6.cob:21: error: LOCAL-STORAGE not allowed in nested programs
```

Comment:
The current behavior is that LOCAL-STORAGE is forbidden in nested programs.
Indeed the standard does not seem to have such restriction...

Reply:
Hm, Roger added that restriction in 2008 because it did not work as expected - OpenCOBOL and later seems to be the only one with that restriction.

It would be a good thing to fix that in a separate commit (maybe it is like the sigsegv that was included here - and it just works correctly when removing that check in parser.y [no need for a dialect option]). Not sure why this is referenced in this PR in the first place...

Reply:
Not sure either. My guess is that when Emilien wrote the profiling tests, he discovered some bugs and made specific tests for them...